streams/shared/src/main/scala/zio/stream/ZPipeline.scala

```diff
@@ -1798,6 +1817,49 @@ object ZPipeline extends ZPipelinePlatformSpecificConstructors {
     new ZPipeline(loop(Chunk.ChunkIterator.empty, 0))
   }
 
+  /**
+   * Creates a pipeline that maps over elements of the stream with the specified
+   * effectful function.
+   *
+   * Unlike `mapZIO` processing is done chunk by chunk. This means that
+   * `mapZIOChunked` provides weaker guarantees than `mapZIO`. While
+   * `stream.mapZIO(f).mapZIO(g)` is guaranteed to be equivalent to
+   * `stream.mapZIO(x => f(x).flatMap(g))`, the same is not true for
+   * `mapZIOChunked`. For example, `mapZIO` guarantees that the first element of
+   * a stream will first be processed with `f` and then `g` before the second
+   * element is processed with `f`. `mapZIOChunked` may process the first two
+   * elements with `f` and only then move on to process the first element with
+   * `g`.
+   */
+  def mapZIOChunked[Env, Err, In, Out](
+    f: In => ZIO[Env, Err, Out]
+  )(implicit trace: Trace): ZPipeline[Env, Err, In, Out] = {
+    def writeWithNext(
+      builder: ChunkBuilder[Out],
+      next: ZChannel[Env, Err, Chunk[In], Any, Err, Chunk[Out], Any]
+    ): ZChannel[Env, Err, Chunk[In], Any, Err, Chunk[Out], Any] = {
+      val out = builder.result()
+      if (out.nonEmpty) ZChannel.write(out) *> next else next
+    }
+
+    lazy val reader: ZChannel[Env, Err, Chunk[In], Any, Err, Chunk[Out], Any] =
+      ZChannel.readWithCause(
+        chunk =>
+          ZChannel.unwrap {
+            val builder = ChunkBuilder.make[Out](chunk.size)
```

Comment:
As a side-note: I wonder if we could reuse the ChunkBuilder after calling `result()` on it but maybe we can re-visit that in a followup PR.

Reply:
I looked into this a bit and I think it is not going to help. The thing is, as long as the stream is not failing or ending, we give the chunk builder a 'sizeHint' that is spot-on; the backing array will be filled to exactlty that size. You should know that if the `ArrayBuilder` fills up the array completely, it will hand out that array in `result` and then builds a new array when the builder is re-used. In other words, the array is always instantiated anew and we might as well also construct a new `ChunkBuilder` so that we get the small benefit of using memory that is local to the CPU core that this is running on.